domain/info.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package domain
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/ddl"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/owner"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/printer"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+const (
+	// ServerInformationPath store server information such as IP, port and so on.
+	ServerInformationPath = "/tidb/server/info"
+	// keyOpDefaultRetryCnt is the default retry count for etcd store.
+	keyOpDefaultRetryCnt = 2
+	// keyOpDefaultTimeout is the default time out for etcd store.
+	keyOpDefaultTimeout = 1 * time.Second
+)
+
+// InfoSessionTTL is the etcd session's TTL in seconds. It's exported for testing.
+var InfoSessionTTL = 10 * 60
+
+// InfoSyncer stores server info to Etcd when when the tidb-server starts and delete when tidb-server shuts down.
+type InfoSyncer struct {
+	etcdCli        *clientv3.Client
+	info           *ServerInfo
+	serverInfoPath string
+	session        *concurrency.Session
+}
+
+// ServerInfo is server static information.
+// It will not update when server running. So please only put static information in ServerInfo struct.
+type ServerInfo struct {
+	ServerVersionInfo
+	ID         string `json:"ddl_id"`
+	IP         string `json:"ip"`
+	Port       uint   `json:"listening_port"`
+	StatusPort uint   `json:"status_port"`
+	Lease      string `json:"lease"`
+}
+
+// ServerVersionInfo is the server version and git_hash.
+type ServerVersionInfo struct {
+	Version string `json:"version"`
+	GitHash string `json:"git_hash"`
+}
+
+// NewInfoSyncer return new InfoSyncer. It is exported for testing.
+func NewInfoSyncer(id string, etcdCli *clientv3.Client) *InfoSyncer {
+	return &InfoSyncer{
+		etcdCli:        etcdCli,
+		info:           getServerInfo(id),
+		serverInfoPath: fmt.Sprintf("%s/%s", ServerInformationPath, id),
+	}
+}
+
+// Init creates a new etcd session and stores server info to etcd.
+func (is *InfoSyncer) Init(ctx context.Context) error {
+	return errors.Trace(is.newSessionAndStoreServerInfo(ctx, owner.NewSessionDefaultRetryCnt))
+}
+
+//GetServerInfo gets self server static information.
+func (is *InfoSyncer) GetServerInfo() *ServerInfo {
+	return is.info
+}
+
+// GetServerInfoByID gets server static information from Etcd.
```

Comment:
We'd better unify the `Etcd`. Some places use "etcd" and others use "Etcd". I think "etcd" is better.